Add exhaustion to damage source and add some javadocs

parlough opened the pull request:
**SpongeAPI** | [SpongeCommon](https://github.com/SpongePowered/SpongeCommon/pull/1573)

Adds the ability to set and retrieve the exhaustion of a damage source. Also adds some Javadocs while I'm at it. I'm open to changes of the defaults for exhaustion and where I set the exhaustion when building the damage sources(refer to common PR).

parlough committed:
Adds exhaustion to damage sources, cleans up javadocs - Fixes #1592

dualspiral commented on src/main/java/org/spongepowered/api/event/cause/entity/damage/source/DamageSource.java:
```diff
@@ -108,24 +108,99 @@ static Builder builder() {
+        /**
+         * Sets the {@link DamageType} of this source.
+         *
+         * <p>This is often required to be set.</p>
```
Often required? When is it not required?

parlough replied:
I thought I found one case where it was not required in some subclass or something, I will look again later though to confirm. I may change wording to "generally".

dualspiral replied:
I would actually avoid words like “generally”, because the problem is that the onus is on the developer to find out when it is and isn’t required. Better to say that the type is required, except in the cases when - if the subclass is a subclass of this builder, the javadoc on the method _there_ should state that it’s not required.

parlough replied:
Fair enough. Best to say when it is not true when it mostly is I suppose. Will work on something more clear.

dualspiral commented on src/main/java/org/spongepowered/api/event/cause/entity/damage/source/DamageSource.java:
```diff
@@ -108,24 +108,99 @@ static Builder builder() {
+         * Sets the amount of exhaustion this {@link DamageSource} will
+         * add to the entity, generally only to players.
+         *
+         * <p>In vanilla this defaults .1 in vanilla, and turns to 0 if the
```
The javadoc here needs rewording a bit, maybe something like:

```
In vanilla gameplay, the default is 0.1, unless the damage is absolute, 
where the exhaustion is zero. This may not be the case for this builder, 
in particular, a call to {@link #absolute()} is not guaranteed to set this to zero.
```

With that being said, I don’t know if that is the best behaviour - I feel like the implementation should follow vanilla mechanics if the exhaustion isn’t set.

parlough replied:
I like this suggestion as well. I would still use the "not guaranteed" wording though, as a source or builder implementing this is not guaranteed to act that way.

Cybermaxke replied:
Bypassing the armor (`bypassesArmor`) should also set the exhaustion to zero.

dualspiral commented on src/main/java/org/spongepowered/api/event/cause/entity/damage/source/DamageSource.java:
```diff
@@ -108,24 +108,99 @@ static Builder builder() {
+
+        /**
+         * Sets this {@link DamageSource} as not being able to be modified
+         * with damage that is absolute.
```
This is a bit confusing - do you actually mean that absolute damage cannot modify this?

What I imagine you mean is that this damage source will be unaffacted by damage modifiers, right?

parlough replied:
I copied this from the get method. I will clean both of them up after looking at how we exactly handle a damage source being absolute in common.

dualspiral replied:
Noted, may as well take advantage of this PR to do that, I guess.

parlough replied:
@dualspiral @gabizou So looking at it now(since we handle damage completely ourselves for the most part), it seems(not 100% sure) we only ignore armor and potion resistance if damage is absolute(in SpongeCommon), but not enchantments, shields, and absorption. Is this an oversight and should we bypass those protections as well, or should I document absolute as just bypassing armor and potion resistance?

parlough replied:
Looks like we were ignoring armor if it was absolute instead of if it was bypassing armor, and were not bypassing enchantments if it was absolute. I will make these changes in my SpongeCommon PR.

To summarize: If a damage source is absolute, like the mcp Javadocs say, it seems to just bypass potions and enchantments.

dualspiral commented on src/main/java/org/spongepowered/api/event/cause/entity/damage/source/common/AbstractDamageSourceBuilder.java:
```diff
@@ -39,9 +39,9 @@
     protected boolean absolute = false;
     protected boolean magical = false;
     protected boolean creative = false;
+    protected double exhaustion = 0.1;
```
...and then I find the impl in the API.

What about making this a nullable `Double`, which indicates “follow vanilla”? Then, when building, this would indicate “do as Vanilla does”.

Just a thought, quite happy to be told this isn’t viable.

parlough replied:
I like your suggestion, I really couldn't decide the best way to handle this. I would be happy to make this change soon/tomorrow. 

I'm definitely open for more feedback here from anyone as this was the hardest part of the change to decide.

_It is unfortunate a lot of this is in the API, but that's a story for later._

dualspiral reviewed (commented):
Few questions on the API, if you don’t mind! :)

parlough committed:
Address concerns

Clean up Javadocs more, switch to following vanilla behavior for exhaustion setting, but allowing them to override it, and address other issues such as wording

parlough commented:
Issues should be addressed, still looking for feedback(and more 3rd-party testing) here and on the common PR.
